lib/compiler/passes/generate-bytecode.js

```diff
@@ -197,18 +196,42 @@ const visitor = require("../visitor");
 //
 //        silentFails--;
 function generateBytecode(ast) {
-  const consts = [];
+  const literals = [];
+  const classes = [];
+  const expectations = [];
+  const functions = [];
 
-  function addConst(value) {
-    const index = consts.indexOf(value);
+  function addLiteralConst(value) {
+    const index = literals.indexOf(value);
 
-    return index === -1 ? consts.push(value) - 1 : index;
+    return index === -1 ? literals.push(value) - 1 : index;
   }
 
-  function addFunctionConst(params, code) {
-    return addConst(
-      "function(" + params.join(", ") + ") {" + code + "}"
-    );
+  function addClassConst(node) {
+    const cls = {
+      value: node.parts,
+      inverted: node.inverted,
+      ignoreCase: node.ignoreCase
+    };
+    const pattern = JSON.stringify(cls);
+    const index = classes.findIndex(c => JSON.stringify(c) === pattern);
```

Comment:
What if classes was a object mapping JSON -> cls?  That way you can do an O(1) lookup and not have to stringify all of the keys on each push.

Reply:
> That way you can do an O(1) lookup

`o(1)` lookup in a Javascript object is generally radically slower than `o(lg n)` lookup into arrays with binary search.  The performance constants and lack of cache coherency typically mean you need an array north of 20,000 elements before the object can pay out.  (it's even worse with `Map`)

Order of complexity is typically wrong in practice.  In the real world, doing it the stupid way with arrays is almost always both faster and smaller, due to the nature of the physical hardware

For a deep discussion of reasoning, turn to Abrash's discussion of why all 3d engines do z-depth sorting with bubble sort, from [Graphics Programming Black Book](https://github.com/jagregory/abrash-black-book), any discussion of why [Judy Arrays tend to pants things with better complexities like Merkle-Damgard trees](http://judy.sourceforge.net/doc/shop_interm.pdf), or just say the words "[cache oblivious datastructure](http://erikdemaine.org/papers/BRICS2002/paper.pdf)" near a Haskell person and hold on to your belt

Typically, big-o should be largely forgotten.  It's mostly a concern for people writing databases

Reply:
We can agree that re-running JSON.stringify for each entry on each insertion is sub-optimal though?

Reply:
I think I agree with @StoneCypher here. Yes, `JSON.stringify` on each insert looks unoptimal, but remember, that
- we make only one pass over the whole AST to fill these arrays
- number of class object hardly ever will exceed even 1000
- using `Map` instead would require some untrivial code to get indexes of entries
  - but we can use a parallel array with stringified objects. `O(n)` search, but without `JSON.stringify` in each insert. I think, benefit from this would too small comparing to the complication of the code